File: Data_structure/Queue/Deque.py
class Node:
    def __init__(self, item):
        self.item = item
        self.l_node = None
        self.r_node = None

class Deque:
    def __init__(self):
        self.cnt = 0
        self.head = None
        self.tail = None

    def size(self):
        return self.cnt

    def empty(self):
        return True if self.cnt == 0 else False

    def front(self):
        if self.empty():
            return -1
        return self.head.item

    def back(self):
        if self.empty():
            return -1
        return self.tail.item

    def push_front(self, item):
        new_node = Node(item)
        if self.head:
            new_node.r_node = self.head
            self.head.l_node = new_node
        else:
            self.tail = new_node
        self.head = new_node
        self.cnt += 1

    def push_back(self, item):
        new_node = Node(item)
        if self.head:
            new_node.l_node = self.tail
            self.tail.r_node = new_node
        else:
            self.head = new_node
        self.tail = new_node
        self.cnt += 1

    def pop_front(self):
        if self.empty():
            return -1
        output = self.front()
        temp = self.head
        self.head = self.head.r_node
        if self.head:
            self.head.l_node = None
        else:
            self.tail = None
        del temp
        self.cnt -= 1
        return output

    def pop_back(self):
        if self.empty():
            return -1
        output = self.back()
        temp = self.tail
        self.tail = self.tail.l_node
        if self.tail:
            self.tail.r_node = None
        else:
            self.head = None
        del temp
        self.cnt -= 1
        return output

    def printall(self):
        node = self.head
        while node:
            print(node.item)
            node = node.r_node

File: Data_structure/Queue/Deque.go
package main

type Deque struct {
	arr []int
}

func NewDeque() *Deque {
	deque := Deque{}
	deque.arr = []int{}
	return &deque
}

func (d Deque) front() int {
	return d.arr[0]
}

func (d Deque) back() int {
	return d.arr[d.size()-1]
}

func (d Deque) size() int {
	return len(d.arr)

}

func (d *Deque) popFront() (int, bool) {
	if d.size() <= 0 {
		return 0, false
	}
	var output int
	output, d.arr = d.arr[0], d.arr[1:]
	return output, true
}

func (d *Deque) popBack() (int, bool) {
	if d.size() <= 0 {
		return 0, false
	}
	var output int
	d.arr, output = d.arr[:d.size()-2], d.size()-1
	return output, true
}

func (d *Deque) insertFront(n int) {
	d.arr = append([]int{n}, d.arr...)
}

func (d *Deque) insertBack(n int) {
	d.arr = append(d.arr, n)
}

func (d Deque) findIndex(n int) int {
	output := -1
	for idx, item := range d.arr {
		if item == n {
			return idx
		}
	}
	return output
}

func (d *Deque) rotateLeft(n int) {
	var left, right []int
	left, right = d.arr[:n], d.arr[n:]
	d.arr = append(right, left...)
}

func (d *Deque) rotateRight(n int) {
	var left, right []int
	left, right = d.arr[:d.size()-n], d.arr[d.size()-n:]
	d.arr = append(right, left...)
}

func main() {
	d := NewDeque()

	d.insertBack(10)
	d.insertBack(9)
	d.insertBack(8)

	print(d.findIndex(8))

	d.insertFront(10)
	d.insertFront(9)
	d.insertFront(8)

	d.rotateLeft(2)
	d.rotateRight(3)

	print()
}
